Add response type factory for specific responses. Allows to append data to jwt. Abstract renderer.

frederikbosch opened the pull request:
Because I was struggling how to properly inject things to a response type, I came to the conclusion that this is not possible in current setup. There is a layer missing. I created `ResponseFactoryInterface`. This solving all kinds of problems.

1. JwtToken can be extended easily
2. ResponseType parameter in the method `respondToRequest` of a grant was only used in case of a BearTokenResponse. Now it takes a factory so it can create the response type that is needed.
3. The ResponseType implementations now only implement the methods required by the interface.
4. Less getters and setters with response types: no more `setAccessToken` for a response type that does not require an access token
5. An access token is not supposed to know anything of a JWT. Only the converter now knows both (because it is a converter).

This is quite a large PR, but I hope you can take the time to see if it beneficial to the library. I think it is.

frederikbosch commented:
Oh, and I also merged PR #501 by @juliangut because it helps solving conflicts beforehand.

warlock39 commented on src/AccessTokenToJwtConverter.php:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @return string
+     */
+    public function convert(AccessTokenEntityInterface $accessTokenEntity)
```
How does it allow to append custom data to JWT? As i can see, it doesn't. Or?

frederikbosch commented:
@warlock39 The PR is updated. The things that you addressed have been taken care of.

1. You can now append data to a JWT token by using a decorator
2. You can now inject a `NullRenderer` to the `ResponseFactory`
3. You can now replace the used JWT implementation with your own by creating your own implementation of `ResponseFactoryInterface`

Regarding making the refresh token optional: that was not part of this PR. The idea of this PR is to give more flexibility with regard to the renderer and resulting token.

frederikbosch commented:
@warlock39 I hope you can give the same kind of feedback again. That would be appreciated since it helps getting better results!

frederikbosch commented:
@warlock39 Did some more commits. Could you have a look?

frederikbosch commented:
Maybe @juliangut can review too, since he has done quite some work too for this library. Hope he does not mind embedding him in this message.. :)

alexbilbie commented:
@frederikbosch if you haven't already could you please update the relevant examples so I can get a better idea of how the final API looks? I'm hoping to be able to review all of the open PRs and issues over the weekend

warlock39 commented on src/Grant/AuthCodeGrant.php:
```diff
@@ -137,59 +145,30 @@ protected function respondToAuthorizationRequest(
-            $htmlResponse->setHtml($html);
-
-            return $htmlResponse;
+            return $responseFactory->newHtmlLoginResponse(
```
Does it allow to use JSON representation as response? Or XML? This method forces me to render html on backend side. It would be great to use strategy here or so. API could look like:
```
return $responseFactory->newLoginResponse(
    LoginData($loginError, $postbackUri, $request->getQueryParams())
);
```

frederikbosch replied:
No, I am going to change more. I had a discussion with @Bobselp already in #499. The `ResponseFactoryInterface` should act as middleware for those people that do not want templating.

frederikbosch replied:
@warlock39 So I am going to extract more of the `AuthCodeGrant` to the `ResponseFactory`. I am going to take your suggestion into account.

warlock39 commented on src/Grant/AuthCodeGrant.php:
```diff
@@ -137,59 +145,30 @@ protected function respondToAuthorizationRequest(
+            );
+
+            return $responseFactory->newHtmlAuthorizeResponse(
+                new AuthorizeData($client, $scopes, $postbackUri, $request->getQueryParams(), $encryptedUserId)
```
To reduce number of arguments (and copy-pastes) `$postbackUri` can be moved to `ResponseFactory` and instead of passing `$request->getQueryParams()` we can pass just `$request`

frederikbosch replied:
Good suggestion, will change.

warlock39 commented on src/Jwt/BearerRedirectResponse.php:
```diff
@@ -0,0 +1,93 @@
+     */
+    public function generateHttpResponse(ResponseInterface $response)
+    {
+        $jwtAccessToken = $this->accessTokenToJwtConverter->convert($this->accessToken)->getToken();
```
`getToken()` is the method of `Lcobucci\JWT\Builder`, so it should not be exposed to client code (because `Builder` is internal dependency of `AccessTokenConverter`)

frederikbosch replied:
Requires another interface, but will change it.

warlock39 commented on src/ResponseTypes/Dto/AuthorizeData.php:
```diff
@@ -0,0 +1,90 @@
+<?php
```
I am not sure, but it seems DTO is redundant

frederikbosch replied:
What do you mean? The namespace?

warlock39 commented:
@frederikbosch commented PR. Not so global, but i hope it will help

frederikbosch commented:
@alexbilbie Have been working on this PR today. To summarize what I have been doing.

1. Only pass a `ServerRequestInterface` and a `DateInterval` to a grant type. It should a return `ResponseTypeInterface`, not also receive one as parameter.
2. I extracted some code from the grant to a `ResponseFactoryInterface`. A response factory is injected into the grant type and could be seen as an object that shares code among the grant types that is related to the response.
3. A default `ResponseFactory` is included and receives as `RendererInterface` and a `AccessTokenConverterInterface`. This encapsulates the default response behaviour as you have defined it in the library: jwt for tokens and html for auth and implicit codes. One can inject its own renderer or access token converter. JWT and the corresponding library `lcobucci/jwt` is abstracted this way.
4. Someone that does not use a renderer, e.g. because only password grant is used, can inject a `NullRenderer` to the default `ResponseFactory`.
5. Someone that does not want html but rather returns `json` to auth code or implicit grants, because they use a SPA, should create their own implementation of `GrantTypeInterface`. It took me a while to think about this. But it is logical. `AuthCodeGrant` is *implementation* of a grant type so it defines how to process a request. If you want to define other behaviour, you should change the implementation. Hence @Bobselp and @warlock39 should create their own `GrantTypeInterface` implementations for their json responses.
6. There is no more `CryptTrait`. There is now a `MessageEncryptionInterface`. This allows people to their own encryption mechanism and/or own encryption library, e.g. `libsodium`.
7. `BearerTokenResponse` and `BearerRedirectResponse` are now unaware of a JWT token. They receive a access token converter.

I also changed the examples as you have requested. However, I did not test the examples. Hopefully you like what I have been doing. Of course, I am open for changes and discussion.

alexbilbie commented:
@frederikbosch is there any way this can be split up into smaller PRs please? This PR is changing way too much for me to try and get my head around all of it
